carmajava/signal_plugin/src/main/java/gov/dot/fhwa/saxton/carma/signal_plugin/TrafficSignalPlugin.java

```diff
@@ -192,40 +196,80 @@ public void onInitialize() {
-                        response.setSuccess(false);
+
+                // If user acknowledged YES that it is safe to continue.
+                if (request.getData()) {
+
+                    log.info("User confirmed all clear.");
+                    log.info("GO_BUTTON_SRVS 3: awaitingUserConfirmation: " + awaitingUserConfirmation + "; awaitingUserInput:" + awaitingUserInput);
+
+                    if (checkCurrentPhase(SignalPhase.GREEN)) { // If still in the GREEN phase
+
+                        //Set false to acknowledge receipt
                         awaitingUserInput.set(false);
+                        awaitingUserConfirmation.set(false);
+
+                        // Notify the arbitrator we need to replan to continue through the intersection
+                        log.info("Current Phase is GREEN after user confirmed. Continuing through intersection and trigger new plan.");
+                        log.info("GO_BUTTON_SRVS 4: awaitingUserConfirmation: " + awaitingUserConfirmation + "; awaitingUserInput:" + awaitingUserInput);
+
+                        // Update vehicle position with most recent transform
+                        updateCurrentLocation(); //INSTEAD OF directly calling triggerNewPlan(true);
+
+                        response.setSuccess(true);
+                        return;
+
+                    } else {//Else no longer green.
+
+                        log.info("Current Phase is NO LONGER GREEN after user confirmed");
+                        log.info("GO_BUTTON_SRVS 5: awaitingUserConfirmation: " + awaitingUserConfirmation + "; awaitingUserInput:" + awaitingUserInput);
+
+                        response.setMessage("No longer green. Please wait to respond until the next green light."); //TODO: TEST
+                        response.setSuccess(false);  //Need to set to false in order to leave the popup open instead of success which the popup disappears.
+
+                        //Set true since popup remains open.
+                        awaitingUserInput.set(true);
+                        awaitingUserConfirmation.set(true);
+
                         return;
                     }
-                } else { // User indicated it is not safe to continue
+
+                } else { // Else user indicated it is not safe to continue
+
                     log.info("User said it is not safe to continue");
-                    prevUIRequestTime.set(System.currentTimeMillis()); // TODO should this be ROS time
+                    response.setMessage("If not safe to continue, please wait for " + (WAIT_INTERVAL/1000) + " seconds...");
+                    response.setSuccess(true); // If we reach this point the service request has been handled
+
+                    // Update the time since user did respond at this time, instead of expiring sooner.
+                    //TODO: should this be ROS time
+                    prevUIRequestTime.set(System.currentTimeMillis());
```

Comment:
Any time which will be compared with timestamps received from external components (pub/sub etc.) needs to use the TimeProvider which can be accessed with the PluginServiceLocator. If it will only be compared internally, it is ok to use System time, but it needs to be consistent do not mix System and TimeProvider.